src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -108,58 +122,95 @@ public PaymentHandleResponse patchAddress(@Nonnull Cart cart, @Nonnull String pa
-                                    format("Can't find cart for interfaceId==[%s]", paypalPlusPaymentId));
-                        }
-                        return patchAddress(cart.get(), paypalPlusPaymentId);
-                    })
-                    .thenApply(paymentHandleResponse -> {
+                                                @Nonnull String paypalPlusPayerId) {
+        CompletionStage<PaymentHandleResponse> executeCS = ctpFacade.getCartService()
+                .getByPaymentMethodAndInterfaceId(PaypalPlusPaymentInterfaceName.PAYPAL_PLUS, paypalPlusPaymentId)
+                .thenCompose(cartOpt -> {
+                    if (!cartOpt.isPresent()) {
+                        return CompletableFuture.completedFuture(PaymentHandleResponse.of404NotFound(
+                                format("Can't find cart for interfaceId==[%s]", paypalPlusPaymentId)));
+                    } else {
+                        Cart cart = cartOpt.get();
+                        PaymentHandleResponse paymentHandleResponse = patchAddress(cart, paypalPlusPaymentId);
                         boolean isSuccessful = HttpStatus.valueOf(paymentHandleResponse.getStatusCode()).is2xxSuccessful();
                         if (isSuccessful) {
                             // execute payment only when patching was successful
-                            return paypalPlusFacade.getPaymentService().execute(new Payment().setId(paypalPlusPaymentId),
-                                    new PaymentExecution().setPayerId(paypalPlusPayerId))
-                                    .thenApply(payment -> PaymentHandleResponse.ofStatusCode(HttpStatus.OK))
-                                    // TODO: re-factor join!!!
-                                    .toCompletableFuture().join();
+                            return setPayerId(paypalPlusPaymentId, paypalPlusPayerId)
+                                    .thenCompose(ctpPayment -> executePaymentAndCreateTxn(paypalPlusPaymentId, paypalPlusPayerId, ctpPayment));
                         } else {
-                            return paymentHandleResponse;
+                            return CompletableFuture.completedFuture(paymentHandleResponse);
                         }
-                    })
-                    .toCompletableFuture().join();
-        } catch (Exception e) {
-            logger.error("Error while processing payment ID {}", paypalPlusPaymentId, e);
-            return PaymentHandleResponse.of500InternalServerError(
-                    format("Error while processing paymentId==[%s]", paypalPlusPaymentId));
-        }
+                    }
+                });
+        return paymentExceptionWrapper(paypalPlusPaymentId, executeCS);
     }
 
-    protected CompletionStage<Payment> updateCtpPayment(Payment newPpPayment, String ctpPaymentId) {
+    protected CompletionStage<io.sphere.sdk.payments.Payment> setPayerId(String paypalPlusPaymentId, String payerId) {
```

Comment:
`updatePaymentWithPayerId`? Otherwise it might be confused with PayPal SDK: `new PaymentExecution().setPayerId(paypalPlusPayerId))`